libass/ass_render.c

```diff
@@ -975,24 +975,37 @@ static void init_font_scale(ASS_Renderer *render_priv)
 {
     ASS_Settings *settings_priv = &render_priv->settings;
 
+    double font_scr_w = render_priv->orig_width;
     double font_scr_h = render_priv->orig_height;
-    if (!render_priv->state.explicit && render_priv->settings.use_margins)
+    if (!render_priv->state.explicit && render_priv->settings.use_margins) {
+        font_scr_w = render_priv->fit_width;
         font_scr_h = render_priv->fit_height;
+    }
+
+    render_priv->screen_scale_x = font_scr_w / render_priv->track->PlayResX;
+    render_priv->screen_scale_y = font_scr_h / render_priv->track->PlayResY;
 
-    render_priv->font_scale = font_scr_h / render_priv->track->PlayResY;
     if (settings_priv->storage_height)
         render_priv->blur_scale = font_scr_h / settings_priv->storage_height;
     else
         render_priv->blur_scale = font_scr_h / render_priv->track->PlayResY;
-    if (render_priv->track->ScaledBorderAndShadow)
-        render_priv->border_scale =
-            font_scr_h / render_priv->track->PlayResY;
-    else
-        render_priv->border_scale = render_priv->blur_scale;
+    if (render_priv->track->ScaledBorderAndShadow) {
+        render_priv->border_scale_x = render_priv->screen_scale_x;
+        render_priv->border_scale_y = render_priv->screen_scale_y;
+    } else {
+        if (settings_priv->storage_width)
+            render_priv->border_scale_x =
+                font_scr_w / settings_priv->storage_width;
```

Comment:
I suspect this additionally needs `/ font_scale_x`. Otherwise it seems to include the anamorphic factor in itself, which the other branches don’t. Imagine a simple situation where PlayRes = display res = storage height but storage width is anamorphic.

To test this: anamorphic video, `ScaledBorderAndShadow: no`, storage size set, and
 the ASS contains… uh… borders or shadows?

Reply:
I don't see any difference between anamorphic shadow/border scaling between xy-VSFilter and libass with this patch, comparison screenshots: [645_compare_anamorphic_border-shadow.tar.gz](https://github.com/libass/libass/files/9507579/645_compare_anamorphic_border-shadow.tar.gz)
Since with `SBAS=no` borders and shadows are storage pixels, and those factors are all meant to scale to rendering frame width it does make sense to me to scale borders/shadows also by the anamorphic (de-)squeeze factor.

Reply:
Right, that makes sense. Thanks for the tests!

I think I got lost in my logical labyrinth last night. It started when I got suspicious of `* border_scale_x / font_scale_x`, which I thought was doubly applying the anamorphism, but I failed to notice that this operation is the inverse and the change I ended up proposing in this thread would precisely make it worse!